dbms/src/TestUtils/executorSerializer.cpp

```diff
@@ -0,0 +1,355 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FmtUtils.h>
+#include <Common/TiFlashException.h>
+#include <Flash/Coprocessor/DAGCodec.h>
+#include <Flash/Coprocessor/DAGUtils.h>
+#include <TestUtils/executorSerializer.h>
+#include <tipb/executor.pb.h>
+#include <tipb/expression.pb.h>
+
+namespace DB::tests
+{
+String ExecutorSerializer::serialize(const tipb::DAGRequest * dag_request)
+{
+    assert((dag_request->executors_size() > 0) != dag_request->has_root_executor());
+    if (dag_request->has_root_executor())
+    {
+        serialize(dag_request->root_executor(), 0);
+        return context.buf.toString();
+    }
+    else
+    {
+        FmtBuffer buffer;
+        String prefix;
+        traverseExecutors(dag_request, [this, &prefix](const tipb::Executor & executor) {
+            assert(executor.has_executor_id());
+            context.buf.fmtAppend("{}{}\n", prefix, executor.executor_id());
+            prefix.append(" ");
+            return true;
+        });
+        return buffer.toString();
+    }
+}
+
+String getJoinTypeName(tipb::JoinType tp)
+{
+    switch (tp)
+    {
+    case tipb::JoinType::TypeAntiLeftOuterSemiJoin:
+        return "AntiLeftOuterSemiJoin";
+    case tipb::JoinType::TypeLeftOuterJoin:
+        return "LeftOuterJoin";
+    case tipb::JoinType::TypeRightOuterJoin:
+        return "RightOuterJoin";
+    case tipb::JoinType::TypeLeftOuterSemiJoin:
+        return "LeftOuterSemiJoin";
+    case tipb::JoinType::TypeAntiSemiJoin:
+        return "AntiSemiJoin";
+    case tipb::JoinType::TypeInnerJoin:
+        return "InnerJoin";
+    case tipb::JoinType::TypeSemiJoin:
+        return "SemiJoin";
+    default:
+        throw TiFlashException(fmt::format("Not supported Join type: {}", tp), Errors::Coprocessor::Internal);
+    }
+}
+
+String getJoinExecTypeName(tipb::JoinExecType tp)
+{
+    switch (tp)
+    {
+    case tipb::JoinExecType::TypeHashJoin:
+        return "HashJoin";
+    default:
+        throw TiFlashException(fmt::format("Not supported Join exectution type: {}", tp), Errors::Coprocessor::Internal);
+    }
+}
+
+String getExchangeTypeName(tipb::ExchangeType tp)
+{
+    switch (tp)
+    {
+    case tipb::ExchangeType::Broadcast:
+        return "Broadcast";
+    case tipb::ExchangeType::PassThrough:
+        return "PassThrough";
+    case tipb::ExchangeType::Hash:
+        return "Hash";
+    default:
+        throw TiFlashException(fmt::format("Not supported Exchange type: {}", tp), Errors::Coprocessor::Internal);
+    }
```

Comment:
how about move this function to `DAGUtils`
and reuse in https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/Statistics/ExchangeSenderImpl.cpp#L35-L48